include/nektar_interface/solver_base/partsys_base.hpp

```diff
@@ -0,0 +1,227 @@
+#include <neso_particles.hpp>
+#include <type_traits>
+
+namespace LU = Nektar::LibUtilities;
+namespace SD = Nektar::SpatialDomains;
+
+namespace NESO::Particles {
+
+class PartSysBase {
+
+public:
+  // Some parameter names used in solver config files
+  inline static const std::string NUM_PARTS_TOT_STR = "num_particles_total";
+  inline static const std::string NUM_PARTS_PER_CELL_STR =
+      "num_particles_per_cell";
+  inline static const std::string PART_OUTPUT_FREQ_STR = "particle_output_freq";
+
+  /// Total number of particles in simulation
+  int64_t num_parts_tot;
+
+  /// NESO-Particles ParticleGroup
+  ParticleGroupSharedPtr particle_group;
+  /// Compute target
+  SYCLTargetSharedPtr sycl_target;
+
+  /**
+   * @brief Write particle parameter values to stdout for any parameter.
+   *  @see also report_param()
+   */
+  inline void add_params_report() {
+    std::cout << "Particle settings:" << std::endl;
+    for (auto const &[param_lbl, param_str_val] : this->param_vals_to_report) {
+      std::cout << "  " << param_lbl << ": " << param_str_val << std::endl;
+    }
+    std::cout << "============================================================="
+                 "=========="
+              << std::endl
+              << std::endl;
+  }
+
+  /**
+   * @brief Clear up memory related to the particle system
+   */
+  inline void free() {
+    if (this->h5part_exists) {
+      this->h5part->close();
+    }
+    this->particle_group->free();
+    this->sycl_target->free();
+    this->particle_mesh_interface->free();
+  };
+
+  /**
+   *  @brief Write particle properties to an output file.
+   *
+   *  @param step Time step number.
+   */
+  inline void write(const int step) {
+    if (this->h5part_exists) {
+      if (this->sycl_target->comm_pair.rank_parent == 0) {
+        nprint("Writing particle properties at step", step);
+      }
+      this->h5part->write();
+    } else {
+      if (this->sycl_target->comm_pair.rank_parent == 0) {
+        nprint("Ignoring call to write particle data because an output file "
+               "wasn't set up. init_output() not called?");
+      }
+    }
+  }
+
+protected:
+  /**
+   * Protected constructor to prohibit direct instantiation.
+   *
+   *  @param session Nektar++ session to use for parameters and simulation
+   * specification.
+   *  @param graph Nektar++ MeshGraph on which particles exist.
+   *  @param comm (optional) MPI communicator to use - default MPI_COMM_WORLD.
+   *
+   */
+  PartSysBase(const LU::SessionReaderSharedPtr session,
+              const SD::MeshGraphSharedPtr graph, ParticleSpec particle_spec,
+              MPI_Comm comm = MPI_COMM_WORLD)
+      : session(session), graph(graph), comm(comm), h5part_exists(false),
+        ndim(graph->GetSpaceDimension()) {
+
+    read_params();
+
+    // Create interface between particles and nektar++
+    this->particle_mesh_interface =
+        std::make_shared<ParticleMeshInterface>(graph, 0, this->comm);
```

Comment:
After creating the interface a call to `extend_halos_fixed_offset(0, particle_mesh_interface);` Will grow the halos by a small amount and improve particle transfer performance. Passing 1 instead of 0 would be a larger halo size by default.

Reply:
So - call this with offset=0 by default, but also provide an option to set a positive offset instead?

Reply:
> So - call this with offset=0 by default, but also provide an option to set a positive offset instead?

Sounds good.

Reply:
Thanks, done